File: skills/wechatpay-basic-payment/references/1-商户/示例代码/Java/15-合单支付/UnionH5Prepay.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/merchant/4014931831

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 合单下单-H5
 */
public class UnionH5Prepay {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "POST";
  private static String PATH = "/v3/combine-transactions/h5";

  public static void main(String[] args) {
    // TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/merchant/4013070756
    UnionH5Prepay client = new UnionH5Prepay(
      "YOUR_MCHID",                    // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/merchant/4013070756
      "YOUR_CERT_SERIAL_NO",         // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013053053
      "/path/to/apiclient_key.pem",     // 商户API证书私钥文件路径，本地文件路径
      "YOUR_PUB_KEY_ID",      // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013038816
      "/path/to/wxp_pub.pem"           // 微信支付公钥文件路径，本地文件路径
    );

    UnionAPIv3H5PrepayRequest request = new UnionAPIv3H5PrepayRequest();
    request.combineAppid = "wxYOUR_APPID";
    request.combineOutTradeNo = "1217752501201407033233368018";
    request.combineMchid = "1230000109";
    request.sceneInfo = new UnionH5SceneInfo();
    request.sceneInfo.payerClientIp = "14.23.150.211";
    request.sceneInfo.deviceId = "013467007045764";
    request.sceneInfo.h5Info = new UnionH5Info();
    request.sceneInfo.h5Info.type = "iOS";
    request.sceneInfo.h5Info.appName = "王者荣耀";
    request.sceneInfo.h5Info.appUrl = "https://pay.qq.com";
    request.sceneInfo.h5Info.bundleId = "com.tencent.wzryiOS";
    request.sceneInfo.h5Info.packageName = "com.tencent.tmgp.sgame";
    request.subOrders = new ArrayList<>();
    {
      UnionCommonSubOrder subOrdersItem0 = new UnionCommonSubOrder();
      subOrdersItem0.mchid = "1230000109";
      subOrdersItem0.outTradeNo = "20150806125346";
      subOrdersItem0.amount = new UnionAmountInfo();
      subOrdersItem0.amount.totalAmount = 10L;
      subOrdersItem0.amount.currency = "CNY";
      subOrdersItem0.attach = "深圳分店";
      subOrdersItem0.description = "腾讯充值中心-QQ会员充值";
      subOrdersItem0.goodsTag = "WXG";
      subOrdersItem0.settleInfo = new UnionSettleInfo();
      subOrdersItem0.settleInfo.profitSharing = false;
      request.subOrders.add(subOrdersItem0);
      UnionCommonSubOrder subOrdersItem1 = new UnionCommonSubOrder();
      subOrdersItem1.mchid = "1230000119";
      subOrdersItem1.outTradeNo = "20150806125347";
      subOrdersItem1.amount = new UnionAmountInfo();
      subOrdersItem1.amount.totalAmount = 10L;
      subOrdersItem1.amount.currency = "CNY";
      subOrdersItem1.attach = "广州分店";
      subOrdersItem1.description = "腾讯充值中心-微信充值";
      subOrdersItem1.goodsTag = "WXG";
      subOrdersItem1.settleInfo = new UnionSettleInfo();
      subOrdersItem1.settleInfo.profitSharing = false;
      request.subOrders.add(subOrdersItem1);
    };
    request.timeExpire = "2018-06-08T10:34:56+08:00";
    request.notifyUrl = "https://yourapp.com/notify";
    try {
      UnionAPIv3H5PrepayResponse response = client.run(request);
        // TODO: 请求成功，继续业务逻辑
        System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
        // TODO: 请求失败，根据状态码执行不同的逻辑
        e.printStackTrace();
    }
  }

  public UnionAPIv3H5PrepayResponse run(UnionAPIv3H5PrepayRequest request) {
    String uri = PATH;
    String reqBody = WXPayUtility.toJson(request);

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo,privateKey, METHOD, uri, reqBody));
    reqBuilder.addHeader("Content-Type", "application/json");
    RequestBody requestBody = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), reqBody);
    reqBuilder.method(METHOD, requestBody);
    Request httpRequest = reqBuilder.build();

    // 发送HTTP请求
    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        // 2XX 成功，验证应答签名
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);

        // 从HTTP应答报文构建返回数据
        return WXPayUtility.fromJson(respBody, UnionAPIv3H5PrepayResponse.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public UnionH5Prepay(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class UnionAPIv3H5PrepayRequest {
    @SerializedName("combine_appid")
    public String combineAppid;
  
    @SerializedName("combine_out_trade_no")
    public String combineOutTradeNo;
  
    @SerializedName("combine_mchid")
    public String combineMchid;
  
    @SerializedName("scene_info")
    public UnionH5SceneInfo sceneInfo;
  
    @SerializedName("sub_orders")
    public List<UnionCommonSubOrder> subOrders = new ArrayList<UnionCommonSubOrder>();
  
    @SerializedName("time_expire")
    public String timeExpire;
  
    @SerializedName("notify_url")
    public String notifyUrl;
  }
  
  public static class UnionAPIv3H5PrepayResponse {
    @SerializedName("h5_url")
    public String h5Url;
  }
  
  public static class UnionH5SceneInfo {
    @SerializedName("payer_client_ip")
    public String payerClientIp;
  
    @SerializedName("device_id")
    public String deviceId;
  
    @SerializedName("h5_info")
    public UnionH5Info h5Info;
  }
  
  public static class UnionCommonSubOrder {
    @SerializedName("mchid")
    public String mchid;
  
    @SerializedName("out_trade_no")
    public String outTradeNo;
  
    @SerializedName("amount")
    public UnionAmountInfo amount;
  
    @SerializedName("attach")
    public String attach;
  
    @SerializedName("description")
    public String description;
  
    @SerializedName("goods_tag")
    public String goodsTag;
  
    @SerializedName("settle_info")
    public UnionSettleInfo settleInfo;
  }
  
  public static class UnionH5Info {
    @SerializedName("type")
    public String type;
  
    @SerializedName("app_name")
    public String appName;
  
    @SerializedName("app_url")
    public String appUrl;
  
    @SerializedName("bundle_id")
    public String bundleId;
  
    @SerializedName("package_name")
    public String packageName;
  }
  
  public static class UnionAmountInfo {
    @SerializedName("total_amount")
    public Long totalAmount;
  
    @SerializedName("currency")
    public String currency;
  }
  
  public static class UnionSettleInfo {
    @SerializedName("profit_sharing")
    public Boolean profitSharing;
  }
  
}

File: skills/wechatpay-basic-payment/references/1-商户/示例代码/Go/15-合单支付/UnionH5Prepay.go
package main

import (
	"bytes"
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/merchant/4015119334
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

func main() {
	// TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/merchant/4013070756
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",                 // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/merchant/4013070756
		"YOUR_CERT_SERIAL_NO",      // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013053053
		"/path/to/apiclient_key.pem", // 商户API证书私钥文件路径，本地文件路径
		"YOUR_PUB_KEY_ID",   // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013038816
		"/path/to/wxp_pub.pem",       // 微信支付公钥文件路径，本地文件路径
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &UnionApiv3H5PrepayRequest{
		CombineAppid:      wxpay_utility.String("wxYOUR_APPID"),
		CombineOutTradeNo: wxpay_utility.String("1217752501201407033233368018"),
		CombineMchid:      wxpay_utility.String("1230000109"),
		SceneInfo: &UnionH5SceneInfo{
			PayerClientIp: wxpay_utility.String("14.23.150.211"),
			DeviceId:      wxpay_utility.String("013467007045764"),
			H5Info: &UnionH5Info{
				Type:        wxpay_utility.String("iOS"),
				AppName:     wxpay_utility.String("王者荣耀"),
				AppUrl:      wxpay_utility.String("https://pay.qq.com"),
				BundleId:    wxpay_utility.String("com.tencent.wzryiOS"),
				PackageName: wxpay_utility.String("com.tencent.tmgp.sgame"),
			},
		},
		SubOrders: []UnionCommonSubOrder{
			UnionCommonSubOrder{
				Mchid:      wxpay_utility.String("1230000109"),
				OutTradeNo: wxpay_utility.String("20150806125346"),
				Amount: &UnionAmountInfo{
					TotalAmount: wxpay_utility.Int64(10),
					Currency:    wxpay_utility.String("CNY"),
				},
				Attach:      wxpay_utility.String("深圳分店"),
				Description: wxpay_utility.String("腾讯充值中心-QQ会员充值"),
				GoodsTag:    wxpay_utility.String("WXG"),
				SettleInfo: &UnionSettleInfo{
					ProfitSharing: wxpay_utility.Bool(false),
				},
			},
			UnionCommonSubOrder{
				Mchid:      wxpay_utility.String("1230000119"),
				OutTradeNo: wxpay_utility.String("20150806125347"),
				Amount: &UnionAmountInfo{
					TotalAmount: wxpay_utility.Int64(10),
					Currency:    wxpay_utility.String("CNY"),
				},
				Attach:      wxpay_utility.String("广州分店"),
				Description: wxpay_utility.String("腾讯充值中心-微信充值"),
				GoodsTag:    wxpay_utility.String("WXG"),
				SettleInfo: &UnionSettleInfo{
					ProfitSharing: wxpay_utility.Bool(false),
				},
			},
		},
		TimeExpire: wxpay_utility.Time(time.Now()),
		NotifyUrl:  wxpay_utility.String("https://yourapp.com/notify"),
	}

	response, err := UnionH5Prepay(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		// TODO: 请求失败，根据状态码执行不同的处理
		return
	}

	// TODO: 请求成功，继续业务逻辑
	fmt.Printf("请求成功: %+v\n", response)
}

func UnionH5Prepay(config *wxpay_utility.MchConfig, request *UnionApiv3H5PrepayRequest) (response *UnionApiv3H5PrepayResponse, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "POST"
		path   = "/v3/combine-transactions/h5"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqBody, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}
	httpRequest, err := http.NewRequest(method, reqUrl.String(), bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	httpRequest.Header.Set("Content-Type", "application/json")
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), reqBody)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		// 2XX 成功，验证应答签名
		err = wxpay_utility.ValidateResponse(
			config.WechatPayPublicKeyId(),
			config.WechatPayPublicKey(),
			&httpResponse.Header,
			respBody,
		)
		if err != nil {
			return nil, err
		}
		response := &UnionApiv3H5PrepayResponse{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}

		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(
			httpResponse.StatusCode,
			httpResponse.Header,
			respBody,
		)
	}
}

type UnionApiv3H5PrepayRequest struct {
	CombineAppid      *string               `json:"combine_appid,omitempty"`
	CombineOutTradeNo *string               `json:"combine_out_trade_no,omitempty"`
	CombineMchid      *string               `json:"combine_mchid,omitempty"`
	SceneInfo         *UnionH5SceneInfo     `json:"scene_info,omitempty"`
	SubOrders         []UnionCommonSubOrder `json:"sub_orders,omitempty"`
	TimeExpire        *time.Time            `json:"time_expire,omitempty"`
	NotifyUrl         *string               `json:"notify_url,omitempty"`
}

type UnionApiv3H5PrepayResponse struct {
	H5Url *string `json:"h5_url,omitempty"`
}

type UnionH5SceneInfo struct {
	PayerClientIp *string      `json:"payer_client_ip,omitempty"`
	DeviceId      *string      `json:"device_id,omitempty"`
	H5Info        *UnionH5Info `json:"h5_info,omitempty"`
}

type UnionCommonSubOrder struct {
	Mchid       *string          `json:"mchid,omitempty"`
	OutTradeNo  *string          `json:"out_trade_no,omitempty"`
	Amount      *UnionAmountInfo `json:"amount,omitempty"`
	Attach      *string          `json:"attach,omitempty"`
	Description *string          `json:"description,omitempty"`
	GoodsTag    *string          `json:"goods_tag,omitempty"`
	SettleInfo  *UnionSettleInfo `json:"settle_info,omitempty"`
}

type UnionH5Info struct {
	Type        *string `json:"type,omitempty"`
	AppName     *string `json:"app_name,omitempty"`
	AppUrl      *string `json:"app_url,omitempty"`
	BundleId    *string `json:"bundle_id,omitempty"`
	PackageName *string `json:"package_name,omitempty"`
}

type UnionAmountInfo struct {
	TotalAmount *int64  `json:"total_amount,omitempty"`
	Currency    *string `json:"currency,omitempty"`
}

type UnionSettleInfo struct {
	ProfitSharing *bool `json:"profit_sharing,omitempty"`
}
